[Merged by Bors] - feat(Algebra): inherit cancellation to subobjects

staroperator opened the pull request:
This PR:
1. adds `Is{Left,Right}Cancel{Add,Mul,MulZero}` instances for `{Add,Mul,MulZero}MemClass` etc.
2. removes the duplicated instances for `Submonoid` and `Submodule` which are introduced in #27955 (benchmark results show there isn't large difference with duplicated instances).

Thanks @Vierkantor's help in benchmarks.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

staroperator committed:
add

staroperator committed:
add CancelMulZero

staroperator committed:
fix doc

eric-wieser commented on Mathlib/Algebra/Group/Submonoid/Defs.lean:
```diff
@@ -458,23 +458,6 @@ instance toMonoid {M : Type*} [Monoid M] (S : Submonoid M) : Monoid S := fast_in
 instance toCommMonoid {M} [CommMonoid M] (S : Submonoid M) : CommMonoid S := fast_instance%
   Subtype.coe_injective.commMonoid Subtype.val rfl (fun _ _ => rfl) fun _ _ => rfl
 
-/-- A submonoid of a left cancellative unital magma inherits left cancellation. -/
-@[to_additive
-      "An `AddSubmonoid` of a left cancellative unital additive magma inherits left cancellation."]
-instance isLeftCancelMul [IsLeftCancelMul M] (S : Submonoid M) : IsLeftCancelMul S :=
-  Subtype.coe_injective.isLeftCancelMul Subtype.val fun _ _ => rfl
-
-/-- A submonoid of a right cancellative unital magma inherits right cancellation. -/
-@[to_additive
-      "An `AddSubmonoid` of a right cancellative unital additive magma inherits right
-      cancellation."]
-instance isRightCancelMul [IsRightCancelMul M] (S : Submonoid M) : IsRightCancelMul S :=
-  Subtype.coe_injective.isRightCancelMul Subtype.val fun _ _ => rfl
-
-/-- A submonoid of a cancellative unital magma inherits cancellation. -/
-@[to_additive "An `AddSubmonoid` of a cancellative unital additive magma inherits cancellation."]
-instance isCancelMul [IsCancelMul M] (S : Submonoid M) : IsCancelMul S where
-
```
@Vierkantor or @j-loreaux, do you remember if we discovered it was better to duplicate these instances for MulMemClass and the concrete instantiations? Certainly we still seem to do so for the `Monoid` instance above.

staroperator committed:
revert `Submodule.isCancelAdd`

staroperator committed:
add for `Semiring` and `Subalgebra`

staroperator committed:
fix ident

staroperator committed:
fix docstring

Vierkantor commented:
Benchmarks seem to have become a lot more stable compared to last month. So let's see what they think:

Vierkantor commented:
!bench

Vierkantor reviewed (commented):
I went ahead and added the keyword "subobjects" to the PR title, to make clear where we inherit the cancellation too. Feel free to adjust as you see fit.

I'll run a benchmark as well, to see what happens if we only inherit for the `MulMemClass`es, and not for the concrete subobjects like `Subalgebra` etc.

Vierkantor committed:
Remove `Is[Left,Right]Cancel[Add,Mul][Zero,]` instances that can be inferred automatically

This is intended mostly to be benchmarked, for comparison with #27992 which keeps these instances. Our hypothesis is that this commit will see performance regressions.

staroperator committed:
Merge branch 'master' into more_cancel

staroperator committed:
Merge branch 'master' into more_cancel

staroperator commented:
bors d+

staroperator commented:
Sorry, bors r+
